File: krun/iteration_runners/iterations_runner.java
// XXX nanotime is actually CLOCK_MONOTONIC on linux, not MONOTONIC_RAW!
// https://bugs.openjdk.java.net/browse/JDK-8006942

// All entry points must implement this
interface BaseKrunEntry {
    public abstract void run_iter(int param);
}

class IterationsRunner {
    public static void main(String args[]) throws
        ClassNotFoundException, NoSuchMethodException, InstantiationException, IllegalAccessException, java.lang.reflect.InvocationTargetException {

        if (args.length != 3) { // not 4 since java doesn't put the program name in args[0]
            System.out.println("usage: iterations_runner <benchmark> <# of iterations> <benchmark param>\n");
            System.exit(1);
        }
        String benchmark = args[0];
        int iterations = Integer.parseInt(args[1]);
        int param = Integer.parseInt(args[2]);

        // reflectively call the benchmark's run_iter
        Class<?> cls = Class.forName(benchmark);
        java.lang.reflect.Constructor<?>[] constructors = cls.getDeclaredConstructors();
        assert constructors.length == 1;
        Object instance = constructors[0].newInstance();
        BaseKrunEntry ke = (BaseKrunEntry) instance; // evil

        System.out.print("[");
        // Please, no refelction inside the timed code!
        for (int i = 0; i < iterations; i++) {
            System.err.println("    Execution: " + (i + 1) + "/" + iterations);

            double startTime = System.nanoTime(); // XXX almost monotonic
            ke.run_iter(param);
            double stopTime = System.nanoTime(); // XXX almost monotonic

            double intvl = (stopTime - startTime) / (float) 1000000000; // nanosecs to secs
            System.out.print(intvl + ", ");
        }
        System.out.print("]");

    }
}

File: krun/iteration_runners/iterations_runner.js
// NOTE: you need to provide clock_gettime_monotonic.

/* Javascript has no assert() */
function _bench_assert(condition) {
    if (!condition) {
        throw "Assertion failed";
    }
}

if (this.arguments.length != 3) {
    throw "usage: iterations_runner.js <benchmark> <# of iterations> <benchmark param>";
}

entry_point = this.arguments[0];
n_iters = parseInt(this.arguments[1]);
param = parseInt(this.arguments[2]);

load(entry_point);

print("[");
for (i = 0; i < n_iters; i++) {
	//print("    Execution " + (i + 1) + "/" + n_iters); // XXX needs to got to stderr
	var start_time = clock_gettime_monotonic();
	run_iter(param);
	var stop_time = clock_gettime_monotonic();

	var intvl = stop_time - start_time;
	print(intvl);
	print(", ")
}
print("]");
